lsbatch/man1/bresources.1
.ds ]W %
.ds ]L
.nh
.TH bresources 1 "OpenLava Version 4.0 - Aug 2016"
.br
.SH NAME
\fBbresources\fR - displays information about resource limits of running jobs
.SH SYNOPSIS
.BR
.PP
.PP
\fBbresources\fR [\fB-h\fR | \fB-V\fR] 
.SH DESCRIPTION
.BR
.PP
.PP
Displays resource limit information.
.PP
.SH OPTIONS
.BR
.PP
.TP
\fB-h
\fR
.IP
Prints command usage to stderr and exits. 


.TP
\fB-V
\fR
.IP
Prints OpenLava release version to stderr and exits. 
.PP
.SH DEFAULT OUTPUT
.BR
.PP
Current usage of resource limits are displayed with the following fields:
.PP
\fBLIMIT\fR
.IP
The name of the resource limit.
.PP
\fBQUEUES or PER_QUEUE\fR
.IP
Queues defined in lsb.resources.
.PP
\fBHOSTS or PER_HOST\fR
.IP
Hosts or host groups defined in lsb.resources.
.PP
\fBPROJECTS or PER_PROJECT\fR
.IP
Projects defined in lsb.resources.
.PP
\fBUSERS or PER_USER\fR
.IP
Users or user groups defined in lsb.resources.
.PP
\fBexpand\fR
.IP
Expanded values of consumer. 
If a user group or host group is used in consumer definition, all members are displayed.
If a group contains a subgroup, each member in the subgroup is displayed recursively.
If not operator (~) is used in consumer definition, all members exclude specified
host/user/queue/project are displayed.
.PP
\fBCURRENT LIMIT RESOURCE USAGE\fR
.IP
Current resrouce usage of projects on this limit.
.PP
\fBPROJECT\fR
.IP
Project defined in this limit has started jobs.
.PP
\fBQUEUE\fR
.IP
The jobs are started on the queue.
.PP
\fBUSER\fR
.IP
The jobs are submitted by the user.
.PP
\fBHOST\fR
.IP
The jobs are running on the host.
.PP
\fBSLOTS\fR
.IP
The number of used slots and the maximum number of slots this project/user/host/queue can use.
.PP
\fBSLOTS_PER_PROCESSOR\fR
.IP
The maximum number of slots per processor this project/user/host/queue can use.
.PP
\fBJOBS\fR
.IP
The number of started jobs and the maximum number of jobs this project/user/host/queue can use.
.PP
.SH EXAMPLE
.PP
bresources
.PP
LIMIT  Limit1
.br
HOSTS      : all ~grp1
.br
    expand : host3 host4
.br
USERS      : all ~openlava
.br
    expand : david wlu
.br
SLOTS      : 0
.br
.PP
LIMIT  Limit2
.br
QUEUES     : normal
.br
    expand : normal
.br
PROJECTS   : Proj1 Proj2 Proj3
.br
JOBS       : 10
.br
CURRENT LIMIT RESOURCE USAGE:
.br
    PROJECT     QUEUE       JOBS
.br
    Proj1       normal      5/10
    Proj2       nornal      2/10
.br
.PP
LIMIT  jenkinslimits
.br
QUEUES     : normal owner
.br
    expand : normal owner
.br
PER_USER   : G1
.br
    expand : david wlu james john
.br
SLOTS      : 20 09:00-17:00
.br
SLOTS      : 10 17:00-09:00
.br
CURRENT LIMIT RESOURCE USAGE:
.br
    PROJECT     QUEUE       USER        HOST        SLOTS       JOBS
.br
    -           owner       john       -           5/10        -
.br
    -           normal      john       -           5/10        -
.PP
.SH SEE ALSO
.BR
.PP
.PP
busers(1), bqueues(1), bhosts(1)


NEWS
Release notes for OpenLava versions

Master
========

-- Implement SLOTS_PER_PROCESSOR in lsb.resources.

4.0beta2
============

-- Add Platform MPI wrapper.
-- Use bash for etc/openlava to support Ubuntu.
-- Fixed rusage string parsing bug.
-- Introduce USUSP and SSUSP counters in the fairshare accounts. Previously
   suspended jobs were accounted as running leading to erroneous state
   transition from suspended to pending state.
-- Implement a new lsf.conf variable named LSB_DEFAULT_USER_GROUP="group name".
   If this variable is specified than jobs that do not specify -G at submission
   time will be assigned LSB_DEFAULT_USER_GROUP as group. This group must be
   configured among valid groups in a fairshare queue.
-- In the fairshare scheduling if a user has jobs that enforce a user group
   and jobs that don't, schedule both types of jobs with -G jobs first.
-- Implement a set of fixes related to fairshare job selection which were
   causing the scheduler to slowly dispatch jobs even if slots were available
   and set incomplete pending reasons.
-- Fixed slot based and resource preemption job replay bug. Fixed slot based
   preemption memory leak, and man pages.
-- Implement a slot based queue preemption in which jobs are suspended,
   instead of requeued.
-- If the MBD does not set the pending reasons for a job print:
   "System is unable to set pending reason for job;" in the bjobs -p
   output instead of blank or meaningless string.
-- badmin ckconfig checks the ownership and permission of work/logdir.
-- In bqueues -l output change the keyword RAN to STARTED to reflect
   the number of started jobs and not only ran.
-- lsid prints the 4.0beta2 identification for the current branch.
-- Implement the decay slot factor for HIST_MINUTES.

4.0beta1
========
-- Fix mbatchd core dump that happens if jobs are submitted with -t option
   of bsub and the job is terminated while pending.
-- Added bjobs -o to support custom output fields
-- Added bjobs -cwd to specify the job current working directory
-- Implement MBD_DEDICATED_RESOURCES in lsf.conf. These are host based boolean
   resources that must be defined in lsf.cluster file. A job must explicitly
   specify a dedicated resource to get allocated the host with that resource.
-- Fix a typo in lsb.queues.in, preemption was spelled with double t.
-- Fix compatibility problems when replaying older lsb.events.
-- Fix the XDR error in the badmin qclose command.
-- Add man page for $LSF_ENVDIR/openlava
-- Add a template file $LSF_ENVDIR/limits.sh
-- Fixed sbatchd compile error on RHEL 5 because glibc lacks CPU_COUNT.
-- Fix a fairshare bug in which jobs were not running under the correct
   user group if bsub -G was specified.
-- Deleted warning message if mbatchd cannot find lsb.resources file.
   lsb.resources is optional.
-- Updated lsb.resources man page for supported use case.
-- Fixed binding incorrect cpus to a queue when HOSTS_SHARES is configured
   in lsb.queues.
-- Enlarge the -J option for the job name to be at least 4096 characters.
-- Added support for binding multiple slots to a core by introducing HOSTS_SHARES
   parameter in lsb.queues for specifying the number of shares the queue is
   allocated for particular host/hostgroup. The share represents the proportion of
   the total number of cores on a host that will be allocated to the queue.
-- Updated non-numa cpu binding feature to support binding multiple
   cores to a job.
-- Added slot limit per host for specific queues in lsb.resources.
-- Added lsb.params PREEMPTABLE_RESOURCES parameters. This enable resource
   preemption in MBD If this parameter is configured the queue preemption
   is done based on the specified resources and not by the host slots which is
   the default.
-- Added the blaunch.1 man page in the batch man section.
-- Developed new API ls_getrpid() to get a remote process id of a task
   started by RES knowing the local taskid.
-- Developed new API lsb_launch() to start a components of parallel jobs.
   This API also collects resource usage of the remote tasks and reports
   the to OpenLava so they can be displayed with the bjobs command.
-- Developed the blaunch command, if mpiexec/mpirun is integrated with
   blaunch, it can launch mpi parallel application, or it can launch
   home grown parallel applications. It is a wrapper around the lsb_launch()
   API.
-- In the bootstrap.sh file create the revision number file
   which is going to be installed in the etc directory.
-- Create lsid 4.0dev4 output.

4.0pre3
=======

-- Create lsid 4.0pre3 tag output.

4.0pre2
========

Introduce 4.0dev3 lsid output. The dev indicates that the code
is between tags. 4.0dev3 finishes with 4.0pre3 tag. After which
we will start 4.0dev4 lsid output till we tag 4.0pre4 or we cut
a final 4.0 branch. In this way when building tagged version we
know the exact commit.

4.0pre1
========

-- Fix res/nios hang blocking ol make.
-- Create a new parameter NIOS_MAX_TASKS specifying the maximum number
   of tasks nios can handle.
-- Fix/reimplement btop/bbot. Since the job queue management has undergone
   performance improvements these function had to be coded as well.
-- Fix the mbatchd variable MBD_NO_QSORT_JOBS, which was incorrectly
   set to MBD_QSORT_JOBS in daemons.c
-- Fix formatting bug in bparams.
-- Add badmin hclose/hopen -C "message", and fix parsing error of "badmin
   subcommand <command_options>".
-- Fix backward compatibility bug in which earlier version of OL cannot
   get information from higher version if mbatchd, specifically 3.0/3.1
   bqueues cannot get information from 3.3 bqueues.
-- Implement lsb.resources of job slot limit for each user/project.
-- Implement the time window in which resource limits are performed.
-- Fix a broken btop/bbot.
-- Introduce a new rule in config/Makefile.am which creates a file named
   commit in the LSF_ENVDIR, this file contains the git commit number
   of the current build.
-- Fix user group bug in lsb.resources.

lsbatch/cmd/bresources.c
/*
 * Copyright (C) 2016 Teraproc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsbatch.h"
#include "cmd.h"
#include "lsb.h"

static void
usage(void)
{
    fprintf(stderr, "bresources: [ -h ] [ -V ]\n");
}

static void
print_limits(struct resLimitReply *);

int
main(int argc, char **argv)
{
    int cc;
    struct resLimitReply *limits;

    if (lsb_init(argv[0]) < 0) {
        lsb_perror("lsb_init");
        return -1;
    }

    while ((cc = getopt(argc, argv, "hV")) != EOF) {
        switch (cc) {
            case 'V':
                fputs(_LS_VERSION_, stderr);
                return 0;
            case 'h':
                usage();
                exit(-1);
        }
    }


    limits = lsb_getlimits();
    if (limits == NULL) {
        if (lsberrno == LSBE_NO_ERROR) {
            fprintf(stderr, "\
bresources: No resource limit yet in the system.\n");
            return -1;
        }
        fprintf(stderr, "bresources: %s.\n", lsb_sysmsg());
        return -1;
    }

    print_limits(limits);

    free_resLimits(limits);

    return 0;
}

/* print_groups
 */
static void
print_limits(struct resLimitReply *reply)
{
    int i, j, k;
    int slimit, jlimit;
    float plimit;
    char *name;
    int  inUse;

    for (i = 0; i < reply->numLimits; i++) {
        printf("%s  %s\n", "LIMIT", reply->limits[i].name);

        for (j = 0; j < reply->limits[i].nConsumer; j++) {
            switch (reply->limits[i].consumers[j].consumer) {
                case LIMIT_CONSUMER_QUEUES:
                    name = "QUEUES";
                    break;
                case LIMIT_CONSUMER_PER_QUEUE:
                    name = "PER_QUEUE";
                    break;
                case LIMIT_CONSUMER_PROJECTS:
                    name = "PROJECTS";
                    break;
                case LIMIT_CONSUMER_PER_PROJECT:
                    name = "PER_PROJECT";
                    break;
                case LIMIT_CONSUMER_HOSTS:
                    name = "HOSTS";
                    break;
                case LIMIT_CONSUMER_PER_HOST:
                    name = "PER_HOST";
                    break;
                case LIMIT_CONSUMER_USERS:
                    name = "USERS";
                    break;
                case LIMIT_CONSUMER_PER_USER:
                    name = "PER_USER";
                    break;
                default:
                    usage();
                    exit(-1);
            }
            printf("%-10s : %s\n", name, reply->limits[i].consumers[j].def);
            if (reply->limits[i].consumers[j].consumer != LIMIT_CONSUMER_PROJECTS
                    && reply->limits[i].consumers[j].consumer != LIMIT_CONSUMER_PER_PROJECT)
                printf("    expand : %s\n", reply->limits[i].consumers[j].value);
        }

        slimit = jlimit = plimit = -1;
        for (j = 0; j < reply->limits[i].nRes; j++) {
             switch (reply->limits[i].res[j].res) {
                case LIMIT_RESOURCE_SLOTS:
                    name = "SLOTS";
                    slimit = (int)reply->limits[i].res[j].value;
                    break;
                case LIMIT_RESOURCE_SLOTS_PER_PROCESSOR:
                    name = "SLOTS_PER_PROCESSOR";
                    plimit = reply->limits[i].res[j].value;
                    break;
                case LIMIT_RESOURCE_JOBS:
                    name = "JOBS";
                    jlimit = (int)reply->limits[i].res[j].value;
                    break;
                default:
                    usage();
                    exit(-1);
            }
            if (reply->limits[i].res[j].res == LIMIT_RESOURCE_SLOTS_PER_PROCESSOR)
                printf("%-10s : %.1f %s\n", name, reply->limits[i].res[j].value, reply->limits[i].res[j].windows);
            else
                printf("%-10s : %d %s\n", name, (int)(reply->limits[i].res[j].value), reply->limits[i].res[j].windows);
        }

        if (reply->numUsage > 0
                && (slimit > 0 || jlimit > 0 || plimit > 0)) {
            printf("\n%s\n", "CURRENT LIMIT RESOURCE USAGE:");
            printf("    %-10s  %-10s  %-10s  %-10s  %-10s  %-10s\n",
                   "PROJECT", "QUEUE", "USER", "HOST", "SLOTS", "JOBS");

            inUse = FALSE;
            for (k = 0; k < reply->numUsage; k++) {
                char buf1[64];
                char buf2[64];
                if (strcmp(reply->limits[i].name, reply->usage[k].limitName)
                        || (reply->usage[k].slots <= 0 && reply->usage[k].jobs <= 0))
                    continue;

                if (slimit > 0 || plimit > 0)
                    sprintf(buf1, "%d/%d", (int)reply->usage[k].slots, (int)reply->usage[k].maxSlots);
                else
                    sprintf(buf1, "%s", "-");

                if (jlimit > 0)
                    sprintf(buf2, "%d/%d", (int)reply->usage[k].jobs, (int)reply->usage[k].maxJobs);
                else
                    sprintf(buf2, "%s", "-");

                printf("    %-10s  %-10s  %-10s  %-10s  %-10s  %-10s\n",
                            !strcmp(reply->usage[k].project, "") ? "-" : reply->usage[k].project,
                            !strcmp(reply->usage[k].queue, "") ? "-" : reply->usage[k].queue,
                            !strcmp(reply->usage[k].user, "") ? "-" : reply->usage[k].user,
                            !strcmp(reply->usage[k].host, "") ? "-" : reply->usage[k].host,
                            buf1,
                            buf2);
                inUse = TRUE;
            }
            if (!inUse)
                printf("    %-10s  %-10s  %-10s  %-10s  %-10s  %-10s\n", "-", "-", "-", "-", "-", "-");
        }
        printf("\n\n");
    }
}
